src/apply-networking-linux.py

```diff
@@ -353,18 +375,25 @@ def set_network_config_file(self, network_data, reset_to_dhcp=False):
             ]
 
             dns = []
+            link_ns = ethernets[os_link_name]["nameservers"]["addresses"]
             for service in network["services"]:
                 if str(service["type"]) == "dns":
+                    if service["address"] in link_ns:
```

Comment:
Why are the DNSses for the same network duplicated in the first place? If you make this assumption, then the OpenStack metadata is not to be trusted that is consistent and correct. Same with the network routes.